doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,232 @@
+the user to cast to their desired serialization format downstream.
+
+We'll add a new endpoint to the existing [`base_router`](https://github.com/MaterializeInc/materialize/blob/6e1f4c7352427301d782438d614feafb0f644442/src/environmentd/src/http.rs#L747)
+in `environmentd` with the path: `/api/webhook/:id`. This follows the existing pattern of our other
+two endpoints, whose paths also start with `/api`.
+
+> **_Note_**: The existing Web Sockets API is at `/api/experimental/ws`, we could move the webhooks
+API under the path `/api/experimental/webhook`, if we think there will be significant breaking
+changes in the future.
+
+We'll also add a new API `SessionClient::append_webhook` which is very similar to the existing
+[`insert_rows`](https://github.com/MaterializeInc/materialize/blob/6e1f4c7352427301d782438d614feafb0f644442/src/adapter/src/client.rs#L392-L397).
+This new API will send the captured `:id` from the URL path, the headers and body of the request,
+straight to the coordinator, via a new `Command`. In the coordinator we'll do the following:
+
+1. Spawn a new `tokio::task` to do the work off the main thread.
+2. Map the `id` to a previously created Source.
+3. Remove sensitive header values from the received headers (e.g. [Authorization](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization)).
+4. Decode the body of the request according to the Source's `FORMAT`.
+5. Pack the filtered headers and decoded body into a `Row`.
+6. Commit that `Row` with a new `StorageController::monotonic_append` API.
+7. Respond once the `StorageController` reports the data as committed.
+
+Today we expose a method [`append`](https://github.com/MaterializeInc/materialize/blob/149d2e5bca1f54e94acfe02ab5f5751b9cb2adfc/src/storage-client/src/controller.rs#L403-L406)
+on the `StorageController` trait, I would like to add a new method:
+
+```
+pub async fn monotonic_append(&self, id: GlobalId, updates: Vec<(Row, Diff)>) -> Result<(), StorageError>;
+```
+
+This new method would call the already existing [`append_to_managed_collection`](https://github.com/MaterializeInc/materialize/blob/149d2e5bca1f54e94acfe02ab5f5751b9cb2adfc/src/storage-client/src/controller.rs#L2718)
+and `.await`-ing the function would complete once the data has been successfully committed. I
+believe we should add this new API instead of using the already existing `append` API, because
+`append` requires the coordinator to pick a timestamp, meanwhile `monotonic_append` would have
+storage pick the timestamp. Having storage select the timestamp aligns better with our goal of
+having a storage owned "push source" in the future.
+
+### Request Limits
+[request_limits]: #request_limits
+
+To start we'll aim to support at least 100 requests per second, in other words, 10 sources could
+receive 10 requests per second. We can probably scale further, but this seems like a sensible limit
```

Comment:
We may also need to limit the total number of bytes, probably per request?

Reply:
Good thought, I updated the request limits to specify a max size of 16KiB, and configurable via LaunchDarkly. I picked this number randomly though and am more than happy to change it!

Reply:
That number sgtm. We...need to make sure this limitation is handled by the http server so that malicious connections get cut off mid connection (that probably shouldn't happen in this document or PRs, but is just a general problem).

Reply:
we should also validate that a persist shard can handle 10 qps of non-empty writes. it's currently tuned for more like 1-2 qps

Reply:
Because this isn't SQL we have lots of flexibility here. We can teach the storage controller to batch these up and either serialize them or wait such that it only does 1-2 qps.

Reply:
Opened https://github.com/MaterializeInc/materialize/issues/20213 to track the implementation of Persist benchmarks, and optionally falling back to the strategy of buffering and writing only once per second